File: listings/src/main/java/chapter05/Listing0519.java
package chapter05;

public class Listing0519 {
  public static void main(String[] args) {
    String[] languages = new String[] {"Kotlin", "Scala", "Java", "Groovy"};

    // Without @JvmOverloads: you must pass in all parameters
    ArrayUtils.join(languages, ";", "{", "}");    // Assumes @file:JvmName("ArrayUtils")

    // With @JvmOverloads: overloaded methods
    ArrayUtils.join(languages);                                 // Skips all optional parameters
    ArrayUtils.join(languages, "; ");                   // Skips prefix and suffix
    ArrayUtils.join(languages, "; ", "Array: ");  // Skips suffix
    ArrayUtils.join(languages, "; ", "[", "]");   // Passes in all possible arguments
  }
}

File: listings/src/main/kotlin/chapter05/Listing0519.kt
@file:JvmName("ArrayUtils")
package chapter05

@JvmOverloads  // Triggers generation of overloaded methods in Java bytecode
fun <T> Array<T>.join(delimiter: String = ", ",
                      prefix: String = "",
                      suffix: String = ""): String {
  return this.joinToString(delimiter, prefix, suffix)
}
